File: DSA-1.cpp
#include <iostream>
using namespace std;

class hash_table {
    int HT1[10], HT2[10]; // HT1 -> Linear , HT2 -> Quadratic
    int size = 10;

public:
    hash_table() {
        for (int i = 0; i < size; i++) {
            HT1[i] = -1;
            HT2[i] = -1;
        }
    }

    void insert_linearProbing();
    void search_linearProbing();
    void delete_linearProbing();
    void insert_quadraticProbing();
    void search_quadraticProbing();
    void delete_quadraticProbing();
    void display();
};

void hash_table::insert_linearProbing() {
    int value, address, count = 0;
    cout << "Enter the data to add in hash table (Linear Probing): ";
    cin >> value;
    address = value % size;

    while (HT1[address] != -1 && count < size) {
        address = (address + 1) % size;
        count++;
    }

    if (count == size) {
        cout << "Hash table is full! Cannot insert " << value << endl;
    } else {
        HT1[address] = value;
        cout << "Inserted " << value << " at index " << address << endl;
    }
}

void hash_table::search_linearProbing() {
    int value, address, count = 0;
    cout << "Enter the number to search in Linear Probing: ";
    cin >> value;
    address = value % size;

    while (HT1[address] != value && HT1[address] != -1 && count < size) {
        address = (address + 1) % size;
        count++;
    }

    if (HT1[address] == value) {
        cout << "Found " << value << " at index " << address << " with " << count + 1 << " comparisons.\n";
    } else {
        cout << "Number not found!\n";
    }
}

void hash_table::delete_linearProbing() {
    int value, address, count = 0;
    cout << "Enter the number to delete from Linear Probing: ";
    cin >> value;
    address = value % size;

    while (HT1[address] != value && HT1[address] != -1 && count < size) {
        address = (address + 1) % size;
        count++;
    }

    if (HT1[address] == value) {
        HT1[address] = -1;
        cout << "Deleted " << value << " from index " << address << endl;
    } else {
        cout << "Number not found!\n";
    }
}

void hash_table::insert_quadraticProbing() {
    int value, address, i = 0;
    cout << "Enter the data to add in hash table (Quadratic Probing): ";
    cin >> value;
    address = value % size;

    while (HT2[(address + i * i) % size] != -1 && i < size) {
        i++;
    }

    if (i == size) {
        cout << "Hash table is full! Cannot insert " << value << endl;
    } else {
        HT2[(address + i * i) % size] = value;
        cout << "Inserted " << value << " at index " << (address + i * i) % size << endl;
    }
}

void hash_table::search_quadraticProbing() {
    int value, address, i = 0, comparisons = 0;
    cout << "Enter the number to search in Quadratic Probing: ";
    cin >> value;
    address = value % size;

    while (HT2[(address + i * i) % size] != value && HT2[(address + i * i) % size] != -1 && i < size) {
        i++;
        comparisons++;
    }

    if (HT2[(address + i * i) % size] == value) {
        cout << "Found " << value << " at index " << (address + i * i) % size << " with " << comparisons + 1 << " comparisons.\n";
    } else {
        cout << "Number not found!\n";
    }
}

void hash_table::delete_quadraticProbing() {
    int value, address, i = 0;
    cout << "Enter the number to delete from Quadratic Probing: ";
    cin >> value;
    address = value % size;

    while (HT2[(address + i * i) % size] != value && HT2[(address + i * i) % size] != -1 && i < size) {
        i++;
    }

    if (HT2[(address + i * i) % size] == value) {
        HT2[(address + i * i) % size] = -1;
        cout << "Deleted " << value << " from index " << (address + i * i) % size << endl;
    } else {
        cout << "Number not found!\n";
    }
}

void hash_table::display() {
    cout << "\nHash Table (Linear Probing):\n";
    for (int i = 0; i < size; i++) {
        cout << i << " -> " << (HT1[i] == -1 ? "empty" : to_string(HT1[i])) << endl;
    }

    cout << "\nHash Table (Quadratic Probing):\n";
    for (int i = 0; i < size; i++) {
        cout << i << " -> " << (HT2[i] == -1 ? "empty" : to_string(HT2[i])) << endl;
    }
}

int main() {
    hash_table obj;
    int ch;

    do {
        cout << "\n___MENU___" << endl;
        cout << "1. Insert data using Linear Probing" << endl;
        cout << "2. Search data using Linear Probing" << endl;
        cout << "3. Delete data using Linear Probing" << endl;
        cout << "4. Insert data using Quadratic Probing" << endl;
        cout << "5. Search data using Quadratic Probing" << endl;
        cout << "6. Delete data using Quadratic Probing" << endl;
        cout << "7. Display both hash tables" << endl;
        cout << "8. Exit" << endl;
        cout << "Enter your choice: ";
        cin >> ch;

        switch (ch) {
            case 1: obj.insert_linearProbing(); break;
            case 2: obj.search_linearProbing(); break;
            case 3: obj.delete_linearProbing(); break;
            case 4: obj.insert_quadraticProbing(); break;
            case 5: obj.search_quadraticProbing(); break;
            case 6: obj.delete_quadraticProbing(); break;
            case 7: obj.display(); break;
            case 8: cout << "Exiting program." << endl; break;
            default: cout << "Invalid choice." << endl;
        }
    } while (ch != 8);

    return 0;
}

//OUTPUT:
__MENU___
1. Insert data using Linear Probing
2. Search data using Linear Probing
3. Delete data using Linear Probing
4. Insert data using Quadratic Probing
5. Search data using Quadratic Probing
6. Delete data using Quadratic Probing
7. Display both hash tables
8. Exit
Enter your choice: 1
Enter the data to add in hash table (Linear Probing): 32
Inserted 32 at index 2

___MENU___
1. Insert data using Linear Probing
2. Search data using Linear Probing
3. Delete data using Linear Probing
4. Insert data using Quadratic Probing

5. Search data using Quadratic Probing
6. Delete data using Quadratic Probing
7. Display both hash tables
8. Exit
Enter your choice: 1
Enter the data to add in hash table (Linear Probing): 45
Inserted 45 at index 5

___MENU___
1. Insert data using Linear Probing
2. Search data using Linear Probing
3. Delete data using Linear Probing
4. Insert data using Quadratic Probing
5. Search data using Quadratic Probing
6. Delete data using Quadratic Probing
7. Display both hash tables
8. Exit
Enter your choice: 4
Enter the data to add in hash table (Quadratic Probing): 78
Inserted 78 at index 8

___MENU___
1. Insert data using Linear Probing
2. Search data using Linear Probing
3. Delete data using Linear Probing
4. Insert data using Quadratic Probing
5. Search data using Quadratic Probing
6. Delete data using Quadratic Probing
7. Display both hash tables
8. Exit
Enter your choice: 1

Enter the data to add in hash table (Linear Probing): 12
Inserted 12 at index 3

___MENU___
1. Insert data using Linear Probing
2. Search data using Linear Probing
3. Delete data using Linear Probing
4. Insert data using Quadratic Probing
5. Search data using Quadratic Probing
6. Delete data using Quadratic Probing
7. Display both hash tables
8. Exit
Enter your choice: 2
Enter the number to search in Linear Probing: 45
Found 45 at index 5 with 1 comparisons.

___MENU___
1. Insert data using Linear Probing
2. Search data using Linear Probing
3. Delete data using Linear Probing
4. Insert data using Quadratic Probing
5. Search data using Quadratic Probing
6. Delete data using Quadratic Probing
7. Display both hash tables
8. Exit
Enter your choice: 5
Enter the number to search in Quadratic Probing: 78
Found 78 at index 8 with 1 comparisons.

___MENU___
1. Insert data using Linear Probing

2. Search data using Linear Probing
3. Delete data using Linear Probing
4. Insert data using Quadratic Probing
5. Search data using Quadratic Probing
6. Delete data using Quadratic Probing
7. Display both hash tables
8. Exit
Enter your choice: 3
Enter the number to delete from Linear Probing: 12
Deleted 12 from index 3

___MENU___
1. Insert data using Linear Probing
2. Search data using Linear Probing
3. Delete data using Linear Probing
4. Insert data using Quadratic Probing
5. Search data using Quadratic Probing
6. Delete data using Quadratic Probing
7. Display both hash tables
8. Exit
Enter your choice: 7

Hash Table (Linear Probing):
0 -> empty
1 -> empty
2 -> 32
3 -> empty
4 -> empty
5 -> 45
6 -> empty
7 -> empty

8 -> empty
9 -> empty

Hash Table (Quadratic Probing):
0 -> empty
1 -> empty
2 -> empty
3 -> empty
4 -> empty
5 -> empty
6 -> empty
7 -> empty
8 -> 78
9 -> empty

___MENU___
1. Insert data using Linear Probing
2. Search data using Linear Probing
3. Delete data using Linear Probing
4. Insert data using Quadratic Probing
5. Search data using Quadratic Probing
6. Delete data using Quadratic Probing
7. Display both hash tables
8. Exit
Enter your choice: 8
Exiting program.

...Program finished with exit code 0
Press ENTER to exit console.

File: DSA-1.py
def Hash(value, n):
    return value % n

def linear_probing_insert(H, value, n):
    index = Hash(value, n)
    i = 0
    while H[(index + i) % n] != -1 and i < n:
        i += 1
    if i == n:
        print("Linear probing failed: Table full")
    else:
        H[(index + i) % n] = value
        print(f"Inserted phone number {value} at index {(index + i) % n}")

def search_linear_probing(H, value, n):
    index = Hash(value, n)
    i = 0
    while H[(index + i) % n] != value and H[(index + i) % n] != -1 and i < n:
        i += 1
    if H[(index + i) % n] == value:
        print(f"Found phone number {value} at index {(index + i) % n} with {i + 1} comparisons.")
    else:
        print("Phone number not found!")

def delete_linear_probing(H, value, n):
    index = Hash(value, n)
    i = 0
    while H[(index + i) % n] != value and H[(index + i) % n] != -1 and i < n:
        i += 1
    if H[(index + i) % n] == value:
        H[(index + i) % n] = -1
        print(f"Deleted phone number {value} from index {(index + i) % n}")
    else:
        print("Phone number not found!")

def quadratic_probing_insert(H, value, n):
    index = Hash(value, n)
    i = 0
    while i < n and H[(index + i * i) % n] != -1:
        i += 1
    if i == n:
        print("Quadratic probing failed: Table full")
    else:
        H[(index + i * i) % n] = value
        print(f"Inserted phone number {value} at index {(index + i * i) % n}")

def search_quadratic_probing(H, value, n):
    index = Hash(value, n)
    i = 0
    while i < n and H[(index + i * i) % n] != value and H[(index + i * i) % n] != -1:
        i += 1
    if H[(index + i * i) % n] == value:
        print(f"Found phone number {value} at index {(index + i * i) % n} with {i + 1} comparisons.")
    else:
        print("Phone number not found!")

def delete_quadratic_probing(H, value, n):
    index = Hash(value, n)
    i = 0
    while i < n and H[(index + i * i) % n] != value and H[(index + i * i) % n] != -1:
        i += 1
    if H[(index + i * i) % n] == value:
        H[(index + i * i) % n] = -1
        print(f"Deleted phone number {value} from index {(index + i * i) % n}")
    else:
        print("Phone number not found!")

def display(H, label):
    print(f"\nHash Table ({label} Probing):")
    for i in range(len(H)):
        print(f"{i} -> {H[i] if H[i] != -1 else 'empty'}")

# --- Main Execution Loop ---
n = int(input("Enter the size of the telephone book (hash table size): "))
H1 = [-1] * n  # Linear Probing Table
H2 = [-1] * n  # Quadratic Probing Table

while True:
    print("\n___TELEPHONE BOOK MENU___")
    print("1. Insert phone number using Linear Probing")
    print("2. Search phone number using Linear Probing")
    print("3. Delete phone number using Linear Probing")
    print("4. Insert phone number using Quadratic Probing")
    print("5. Search phone number using Quadratic Probing")
    print("6. Delete phone number using Quadratic Probing")
    print("7. Display both hash tables")
    print("8. Exit")
    
    choice = int(input("Enter your choice: "))

    if choice == 1:
        x = int(input("Enter phone number to insert: "))
        linear_probing_insert(H1, x, n)
    elif choice == 2:
        x = int(input("Enter phone number to search: "))
        search_linear_probing(H1, x, n)
    elif choice == 3:
        x = int(input("Enter phone number to delete: "))
        delete_linear_probing(H1, x, n)
    elif choice == 4:
        x = int(input("Enter phone number to insert: "))
        quadratic_probing_insert(H2, x, n)
    elif choice == 5:
        x = int(input("Enter phone number to search: "))
        search_quadratic_probing(H2, x, n)
    elif choice == 6:
        x = int(input("Enter phone number to delete: "))
        delete_quadratic_probing(H2, x, n)
    elif choice == 7:
        display(H1, "Linear")
        display(H2, "Quadratic")
    elif choice == 8:
        print("Exiting Telephone Book System.")
        break
    else:
        print("Invalid choice. Please try again.")
